plugins/teststeps/sshcmd/sshcmd.go

```diff
@@ -186,31 +201,39 @@ func (ts *SSHCmd) Run(cancel, pause <-chan struct{}, ch test.TestStepChannels, p
 			errCh <- innerErr
 		}()
 
-		select {
-		case err := <-errCh:
-			log.Infof("Stdout of command '%s' is '%s'", cmd, stdout.Bytes())
-			if err == nil {
-				// Execute expectations
-				expect := ts.Expect.String()
-				if expect == "" {
-					log.Warningf("no expectations specified")
-				} else {
-					re := regexp.MustCompile(expect)
-					matches := re.FindAll(stdout.Bytes(), -1)
-					if len(matches) > 0 {
-						log.Infof("match for regex \"%s\" found", expect)
+		for {
+			select {
+			case err := <-errCh:
+				log.Infof("Stdout of command '%s' is '%s'", cmd, stdoutBuf.Bytes())
+				if err == nil {
+					// Execute expectations
+					expect := ts.Expect.String()
+					if expect == "" {
+						log.Warningf("no expectations specified")
 					} else {
-						return fmt.Errorf("match for %s not found for target %v", expect, target)
+						re := regexp.MustCompile(expect)
+						matches := re.FindAll(stdoutBuf.Bytes(), -1)
+						if len(matches) > 0 {
+							log.Infof("match for regex \"%s\" found", expect)
+						} else {
+							return fmt.Errorf("match for %s not found for target %v", expect, target)
+						}
 					}
+				} else {
+					log.Warningf("Stderr of command '%s' is '%s'", cmd, stderrBuf.Bytes())
 				}
-			} else {
-				log.Warningf("Stderr of command '%s' is '%s'", cmd, stderr.Bytes())
+				return err
+			case <-cancel:
+				return session.Signal(ssh.SIGKILL)
+			case <-pause:
+				return session.Signal(ssh.SIGKILL)
+			default:
+				if time.Now().After(timeTimeout) {
+					return fmt.Errorf("timed out after %s", timeout)
+				}
+				session.Signal(ssh.Signal("CONT"))
+				time.Sleep(250 * time.Millisecond)
```

Comment:
You don't need to busy wait for this, I think the usual way to do this in go is:
```suggestion
		case <-time.After(timeTimeout):
				return fmt.Errorf("timed out after %s", timeout)
				session.Signal(ssh.Signal("SIGKILL"))
```
This should only ever be reached if the first case was not reached before.
(I think you also want to abort execution then with SIGKILL)

Reply:
I know that we do not need busy waiting for this. This actually comes from https://github.com/facebookincubator/contest/issues/152 - I had the problem that the other connection closed the connection. Turned out that I constantly need to fire Signals in order to keep the connection alive. I tried implementing keep alive which did not help - so I just this.

Reply:
Aaaah, that is what the CONT is for. Please drop a comment in then, so people know :)

Reply:
Wait, but why does this help against the connection being closed explicitly? Sry, not sure I follow :/
And 4 signals per second seems like a lot :D